README.md
## What is this ?
This is a simple lightweight **JS Time Ago module**,
that supports different languges and you can contribute to add yours.

## works on the browser and server
if you want to use it on the **browser**, you must include the `nta.min.js` file into you html file
<br/>
else if you are using **server side rendering or any js framework**, just npm install it and import it as you import any other package 


## Languages available for now are:
`ar: arabic, en: english, de: german, vn: vietnamese`

## Display formats

```javascript
// ago format, this is the default format
timeAgo('2020-01-16 19:48:45', 'en'); // will return: 3mo ago

// full format
timeAgo('2020-01-01 19:48:45', 'en', 'full'); // will return: 1 January

```


## Usage
after installing `nta`

```javascript
const timeAgo = require('nta');

timeAgo(new Date, 'en', 'ago')
```

***params*** <br/>
timeAgo function accepts 3 parameter wich are the `date, language, format`. <br/>
**the language is not required, arabic is set by default** <br/>
**the format is not required too, "ago" is set by default**


## Accessible to your views ?

```javascript
app.locals.timeAgo = timeAgo
```
Thats is it, nta is now accessible to all of your views you can use it as specified above.


**Dont Forget to contribute by adding your language or any other feature you like**

CHANGE_LOG.md
# Change Log

## `2.0.0`

### Added

- Browser support.
- Tests with the help of Jest.
- default language to arabic and format to "ago"

### Fixed

- Restrtuctured the code base.
- Fixed: better arabic translations.

test/ar.test.js
const timeAgo = require('../nta.min.js');

describe('ar', () => {
    describe('ago format', () => {
        test('Just now', () => {
            var just_now = new Date();
            expect(timeAgo(just_now)).toBe('الآن');
        });

        test('minutes', () => {
            var last_min_date = new Date(new Date() - (60 * 1000));
            expect(timeAgo(last_min_date)).toBe('منذ دقيقة');
        });
        
        test('hours', () => {
            var _20_hrs_date = new Date(new Date() - (60 * 60 * 20 * 1000));
            expect(timeAgo(_20_hrs_date)).toBe('منذ 20 ساعة');
        });

        test('days', () => {
            var _5_days_date = new Date(new Date() - (60 * 60 * 24 * 5 * 1000));
            expect(timeAgo(_5_days_date)).toBe('منذ 5 أيام');
        });

        test('weeks', () => {
            var _3_weeks_date = new Date(new Date() - (60 * 60 * 24 * 7 * 3 * 1000));
            expect(timeAgo(_3_weeks_date)).toBe('منذ 3 اسابيع');
        });

        test('months', () => {
            var _7_months_date = new Date(new Date() - (2629440 * 7 * 1000));
            expect(timeAgo(_7_months_date)).toBe('منذ 7 أشهر');
        });

        test('years', () => {
            var _5_years_date = new Date(new Date() - (31553280 * 5 * 1000));
            expect(timeAgo(_5_years_date)).toBe('منذ 5 سنوات');
        });
        
    });

    // full format
    describe('full format', () => {
        test('full format, arabic language', () => {

            var yesterday_date = new Date(new Date() - (60 * 60 * 24 * 1000));
            expect(timeAgo(yesterday_date, 'ar', 'full')).toBe('البارحة');

            // var last_week_date = new Date(new Date() - (60 * 60 * 24 * 7 * 1000));
            // expect(timeAgo(last_week_date, 'ar', 'full')).toBe(`أفريل ${last_week_date.getDate()}`);

        });
    })
})



test/en.test.js
const timeAgo = require('../nta.min.js');

describe('en', () => {
    describe('ago format', () => {
        test('Just now', () => {
            var just_now = new Date();
            expect(timeAgo(just_now, 'en')).toBe('now');
        });

        test('minutes', () => {
            var last_min_date = new Date(new Date() - (60 * 1000));
            expect(timeAgo(last_min_date, 'en')).toBe('a minute ago');
        });

        test('hours', () => {
            var _20_hrs_date = new Date(new Date() - (60 * 60 * 20 * 1000));
            expect(timeAgo(_20_hrs_date, 'en')).toBe('20hrs ago');
        });

        test('days', () => {
            var _5_days_date = new Date(new Date() - (60 * 60 * 24 * 5 * 1000));
            expect(timeAgo(_5_days_date, 'en')).toBe('5d ago');
        });

        test('weeks', () => {
            var _3_weeks_date = new Date(new Date() - (60 * 60 * 24 * 7 * 3 * 1000));
            expect(timeAgo(_3_weeks_date, 'en')).toBe('3w ago');
        });

        test('months', () => {
            var _7_months_date = new Date(new Date() - (2629440 * 7 * 1000));
            expect(timeAgo(_7_months_date, 'en')).toBe('7mo ago');
        });

        test('years', () => {
            var _5_years_date = new Date(new Date() - (31553280 * 5 * 1000));
            expect(timeAgo(_5_years_date, 'en')).toBe('5 years ago');
        });

    });

    // full format
    describe('full format', () => {
        test('full format', () => {

            var yesterday_date = new Date(new Date() - (60 * 60 * 24 * 1000));
            expect(timeAgo(yesterday_date, 'en', 'full')).toBe('Yesterday');

            // var last_week_date = new Date(new Date() - (60 * 60 * 24 * 7 * 1000));
            // expect(timeAgo(last_week_date, 'ar', 'full')).toBe(`April ${last_week_date.getDate()}`);

        });
    })
})



src/timeAgo.test.js
const timeAgo = require('../nta.js');

// Install the code runner extension for vscode
// and then hit "ctrl + alt + N" to see the result
console.log(timeAgo('2020-01-01 19:48:45', 'vn', 'ago'));

nta.js
"use strict";var langs={ar:{MONTH_NAMES:["جانفي","فيفري","مارس","أفريل","ماي","جوان","جويلية","أوت","سبتمبر","أكتوبر","نوفمبر","ديسمبر"],ago:{now:"الآن",sec:"منذ !t ثواني",onemin:"منذ دقيقة واحدة",min:"منذ !t دقيقة",onehour:"منذ ساعة",hrs:"منذ !t ساعات",oneday:"منذ يوم",days:"منذ !t أيام",oneweek:"منذ أسبوع",weeks:"منذ !t اسابيع",onemonth:"منذ شهر",months:"منذ !t أشهر",oneyear:"منذ عام",years:"منذ !t سنوات",today:"اليوم",yesterday:"البارحة"}},en:{MONTH_NAMES:["January","February","March","April","May","June","July","August","September","October","November","December"],ago:{now:"now",sec:"!ts ago",onemin:"one minute ago",min:"!tmin ago",onehour:"an hour ago",hrs:"!thrs ago",oneday:"a day ago",days:"!td ago",oneweek:"a week ago",weeks:"!tw ago",onemonth:"a month ago",months:"!tmo ago",oneyear:"a year ago",years:"!t years ago",today:"Today",yesterday:"Yesterday"}},de:{MONTH_NAMES:["Januar","Februar","März","April","Mai","Juni","Juli","August","September","Oktober","November","Dezember"],ago:{now:"jetzt",sec:"Vor !t sec",onemin:"vor einer Minute",min:"vor !t Minuten",onehour:"vor einer Stunde",hrs:"vor !t Stunden",oneday:"vor einem Tag",days:"vor !t Tagen",oneweek:"vor einer Woche",weeks:"Vor !t Wochen",onemonth:"vor einem Monat",months:"vor !t Monaten",oneyear:"vor einem Jahr",years:"vor !t Jahren",today:"Heute",yesterday:"Gestern"}},vn:{MONTH_NAMES:["tháng Giêng","Tháng hai","tháng Ba","Tháng 4","có thể","Tháng 6","Tháng 7","tháng Tám","Tháng Chín","Tháng Mười","Tháng 11","Tháng 12"],ago:{now:"hiện nay",sec:"!t giây trước",onemin:"một phút trước",min:"!t phút trước",onehour:"một giờ trước",hrs:"!t giờ trước",oneday:"một ngày trước",days:"!t ngày trước",oneweek:"một tuần trước",weeks:"!t tuần trước",onemonth:"một tháng trước",months:"!t tháng trước",oneyear:"một năm trước",years:"!t năm trước",today:"Hôm nay",yesterday:"Hôm qua"}}};function getFormattedDate(e,t,a,n){var o=e.getDate(),r=langs[t].MONTH_NAMES[e.getMonth()],g=e.getFullYear(),s=e.getMinutes();return s<10&&(s="0"+s),a||(n?o+". "+r:o+". "+r+" "+g)}function fullFormat(e,t){if(!e)return null;var a="object"==typeof e?e:new Date(e),n=new Date,o=new Date(n-864e5),r=Math.round((n-a)/1e3),g=Math.round(r/60),s=n.toDateString()===a.toDateString(),m=o.toDateString()===a.toDateString(),h=n.getFullYear()===a.getFullYear();return r<5?langs[t].ago.now:r<60?langs[t].ago.sec.replace("!t",r):r<90?langs[t].ago.onemin:g<60?langs[t].ago.min.replace("!t",g):s?getFormattedDate(a,t,langs[t].ago.today):m?getFormattedDate(a,t,langs[t].ago.yesterday):h?getFormattedDate(a,t,!1,!0):getFormattedDate(a,t)}function agoFormat(e,t){if(!e)return null;var a=Date.parse(e)/1e3,n=Date.parse(new Date)/1e3-a,o=n,r=Math.round(n/60),g=Math.round(n/3600),s=Math.round(n/86400),m=Math.round(n/604800),h=Math.round(n/2629440),u=Math.round(n/31553280);return o<=60?o<=5?langs[t].ago.now:langs[t].ago.sec.replace("!t",o):r<=60?1==r?langs[t].ago.onemin:langs[t].ago.min.replace("!t",r):g<=24?1==g?langs[t].ago.onehour:langs[t].ago.hrs.replace("!t",g):s<=7?1==s?langs[t].ago.oneday:langs[t].ago.days.replace("!t",s):m<=4.3?1==m?langs[t].ago.oneweek:langs[t].ago.weeks.replace("!t",m):h<=12?1==h?langs[t].ago.onemonth:langs[t].ago.months.replace("!t",h):1==u?langs[t].ago.oneyear:langs[t].ago.years.replace("!t",u)}function timeAgo(e,t,a){return"ago"==a?agoFormat(e,t):fullFormat(e,t)}module.exports=timeAgo;

gulpfile.js
let gulp = require('gulp');
let concat = require('gulp-concat');
let uglify = require('gulp-uglify');
var inject = require('gulp-inject');

gulp.task('build', () => {
    return gulp.src( './src/timeAgo.js')
                .pipe(inject(
                    gulp.src('./src/lang.js', { read: true }),
                    { 
                        starttag: '//- inject:lang:{{ext}}', 
                        endtag: '//- endinject',
                        transform: function (filePath, file) {
                            // return file contents as string
                            return file.contents.toString('utf8')
                        }
                    }
                ))
                .pipe(concat('nta.min.js'))
                .pipe(uglify())
                .pipe(gulp.dest('./'))
});

src/lang.js
/**
 * Time Languages
 * By the way the "!t" word is a constant it will be replaced with ago date so dont touch it
 */
var langs = {
    /* Arabic languge */
    ar: {
        'MONTH_NAMES': [
            'جانفي', 'فيفري', 'مارس', 'أفريل', 'ماي', 'جوان',
            'جويلية', 'أوت', 'سبتمبر', 'أكتوبر', 'نوفمبر', 'ديسمبر'
        ],
        ago: {
            now: 'الآن',
            moments: 'منذ لحظات',
            sec: {
                many: 'منذ !t ثواني', // e.g: منذ 5 ثواني
                single: 'منذ !t ثانية' // e.g: منذ 30 ثانية
            },
            onemin: 'منذ دقيقة',
            min: {
                many: 'منذ !t دقائق', // e.g: قبل 10 دقائق
                single: 'منذ !t دقيقة' // e.g: قبل 15 دقيقة
            },
            onehour: 'منذ ساعة',
            hrs: {
                many: 'منذ !t ساعات', // e.g: منذ 10 ساعات
                single: 'منذ !t ساعة' // e.g: منذ 20 ساعة
            },
            oneday: 'منذ يوم',
            days: {
                many: 'منذ !t أيام', // e.g: 5d ago
                single: 'منذ !t يومًا'
            },
            oneweek: 'منذ أسبوع',
            weeks: {
                many: 'منذ !t اسابيع', //e.g: منذ 5 أسابيع
                single: 'منذ !t أسبوع' //e.g: منذ 11 أسبوع
            },
            onemonth: 'منذ شهر',
            months: {
                many: 'منذ !t أشهر',
                single: 'منذ !t شهرا'
            },
            oneyear: 'منذ عام',
            years: {
                many: 'منذ !t سنوات',
                single: 'منذ !t عاما'
            },
            today: 'اليوم',
            yesterday: 'البارحة'
        }
    },

    /* English languge */
    en: {
        'MONTH_NAMES': [
            "January", "February", "March", "April", "May", "June",
            "July", "August", "September", "October", "November", "December"
        ],
        ago: {
            now: 'now',
            moments: 'moments ago',
            sec: '!ts ago', //e.g: 2s ago
            onemin: 'a minute ago',
            min: '!t min ago', //e.g: 5min ago
            onehour: 'an hour ago',
            hrs: '!thrs ago', // e.g: 2hrs ago
            oneday: 'a day ago',
            days: '!td ago', // e.g: 5d ago
            oneweek: 'a week ago',
            weeks: '!tw ago', //e.g: 5w ago
            onemonth: 'a month ago',
            months: '!tmo ago',
            oneyear: 'a year ago',
            years: '!t years ago',
            today: 'Today',
            yesterday: 'Yesterday'
        }
    },

    /* German languge */
    de: {
        'MONTH_NAMES': [
            "Januar", "Februar", "März", "April",  "Mai", "Juni", 
            "Juli", "August", "September", "Oktober", "November", "Dezember"
        ],
        ago: {
            now: 'jetzt',
            moments: 'Sekunden zuvor',
            sec: 'Vor !t sec',
            onemin: 'vor einer Minute',
            min: 'vor !t Minuten',
            onehour: 'vor einer Stunde',
            hrs: 'vor !t Stunden', // e.g: 2hrs ago
            oneday: 'vor einem Tag',
            days: 'vor !t Tagen', // e.g: 5d ago
            oneweek: 'vor einer Woche',
            weeks: 'Vor !t Wochen', //e.g: 5w ago
            onemonth: 'vor einem Monat',
            months: 'vor !t Monaten',
            oneyear: 'vor einem Jahr',
            years: 'vor !t Jahren',
            today: 'Heute',
            yesterday: 'Gestern'
        }
    },

    /* Vietnamese languge */
    vn: {
        'MONTH_NAMES': [
            "tháng Giêng", "Tháng hai", "tháng Ba", "Tháng 4", "có thể", "Tháng 6",
            "Tháng 7", "tháng Tám", "Tháng Chín", "Tháng Mười", "Tháng 11", "Tháng 12"
        ],
        ago: {
            now: 'hiện nay',
            moments: 'Khoảnh khắc trước',
            sec: '!t giây trước', //e.g: 2s ago
            onemin: 'một phút trước',
            min: '!t phút trước', //e.g: 5min ago
            onehour: 'một giờ trước',
            hrs: '!t giờ trước', // e.g: 2hrs ago
            oneday: 'một ngày trước',
            days: '!t ngày trước', // e.g: 5d ago
            oneweek: 'một tuần trước',
            weeks: '!t tuần trước', //e.g: 5w ago
            onemonth: 'một tháng trước',
            months: '!t tháng trước',
            oneyear: 'một năm trước',
            years: '!t năm trước',
            today: 'Hôm nay',
            yesterday: 'Hôm qua'
        }
    },
};

src/timeAgo.js
/**
 * Time ago module
 * version: 2.0.0
 * (c) 2020 wassim ben jdida, github/wassimbj
 */


(function(global, factory){
    "use strict";

    // import the helpful code based on the environment
    if(typeof module !== 'undefined' && typeof exports === 'object'){
        module.exports = factory(global);
    }
    else if(typeof define === 'function' && define.amd){
        define('nta', factory(global));
    }else{
        window.nta = factory(global);
    }


}(typeof window !== 'undefined' ? window : this, function(){
    "use strict";

    /*
      Very important, dont remove this comment down here, cause when we run the build command, 
      gulp will inject the langs object in the lang.js file here
    */
    //- inject:lang:js
    //- endinject
    // ###################################################
    

    // Helper
    // throw errors
    function throw_error(msg){
        throw new Error(msg);
    }

    // Helper
    // get formatted date for the fullFormat
    function getFormattedDate(date, lang, prefomattedDate, hideYear) {
        var day = date.getDate();
        var month = langs[lang].MONTH_NAMES[date.getMonth()];
        var year = date.getFullYear();
        var minutes = date.getMinutes();

        // Adding leading zero to minutes
        if (minutes < 10) {
            minutes = '0' + minutes;
        }

        // Today
        // Yesterday
        if (prefomattedDate) {
            return prefomattedDate;
        }

        // 10 January
        // return `${day} ${month}`;
        if (hideYear) {
            // Right to Left
            if(lang == 'ar'){
                return month + ' ' + day;
            }
            // else
            return month + ' ' + day;
        }

        // 10 January 2017
        // return `${day}. ${month} ${year}`;
        return '' + day + ' ' + month + ' ' + year + '';
    }

    // Return a full format, (e.g: 14, January 2019)
    function fullFormat(dateToConvert, lang) {
        if (!dateToConvert) {
            return null;
        }

        var date = typeof dateToConvert === 'object' ? dateToConvert : new Date(dateToConvert);
        var DAY_IN_MS = 86400000; // 24(Hours in a day) * 60(Min in Hours) * 60(Secs in Min) * 1000(Ms in Sec)
        var today = new Date();
        var yesterday = new Date(today - DAY_IN_MS);
        var seconds = Math.round((today - date) / 1000);
        var minutes = Math.round(seconds / 60);
        var isToday = today.toDateString() === date.toDateString();
        var isYesterday = yesterday.toDateString() === date.toDateString();
        var isThisYear = today.getFullYear() === date.getFullYear();


        if (seconds < 5) {
            return langs[lang].ago.now;
        } else if (seconds < 60) {
            if (seconds <= 5) {
                return langs[lang].ago.now;
            } else if (seconds >= 5 && seconds <= 20) {
                return langs[lang].ago.moments.replace('!t', seconds);
            } else {
                return langs[lang].ago.sec.single.replace('!t', seconds);
            }
        } else if (seconds > 60 && seconds < 90) {
            return langs[lang].ago.onemin;
        } else if (minutes < 60) {
            // for arabic
            if (lang == 'ar' && minutes >= 11) {
                return langs[lang].ago.min.single.replace('!t', minutes);
            } else if (lang == 'ar' && minutes <= 11) {
                return langs[lang].ago.min.many.replace('!t', minutes);
            }
            // else
            return langs[lang].ago.min.replace('!t', minutes);
        } else if (isToday) {
            return getFormattedDate(date, lang, langs[lang].ago.today); // Today
        } else if (isYesterday) {
            return getFormattedDate(date, lang, langs[lang].ago.yesterday); // Yesterday
        } else if (isThisYear) {
            return getFormattedDate(date, lang, false, true); // 10 January
        }

        return getFormattedDate(date, lang); // 10 January 2017
    }

    // ago format (e.g: 9w ago, 1min ago...)
    function agoFormat(dateToConvert, lang) {

        if (!dateToConvert)
            return null

        var time_ago = Date.parse(dateToConvert) / 1000;
        var current_time = Date.parse(new Date()) / 1000;
        var diff = current_time - time_ago;

        var sec = diff;
        var min = Math.round(diff / 60); // 1min = 60 seconds
        var hours = Math.round(diff / 3600); // 3600 = 60min * 60min
        var day = Math.round(diff / 86400); // 86400 = 24h * 60min * 60min
        var week = Math.round(diff / 604800); // 604800 = 7d * 24h * 60min * 60min
        var month = Math.round(diff / 2629440); // months
        var year = Math.round(diff / 31553280); // years

        // seconds
        if (sec < 60) {
            if (sec <= 5) {
                return langs[lang].ago.now;
            } else if (sec >= 5 && sec <= 20){
                return langs[lang].ago.moments.replace('!t', sec);
            } else {
                return langs[lang].ago.sec.single.replace('!t', sec);
            }
        } 

        // minutes
        else if (min <= 60) {
            if (min == 1){
                return langs[lang].ago.onemin;
            } else{
                // for arabic
                if(lang == 'ar' && min >= 11){
                    return langs[lang].ago.min.single.replace('!t', min);
                }else if(lang == 'ar' && min <= 11){
                    return langs[lang].ago.min.many.replace('!t', min);
                }
                // else
                return langs[lang].ago.min.replace('!t', min);
            }
        } 
        
        // hours
        else if (hours <= 24) {
            if (hours == 1){
                return langs[lang].ago.onehour;
            }else{
                // for arabic
                if (lang == 'ar' && hours >= 11) {
                    return langs[lang].ago.hrs.single.replace('!t', hours);
                } else if (lang == 'ar' && hours <= 11) {
                    return langs[lang].ago.hrs.many.replace('!t', hours);
                }
                // else
                return langs[lang].ago.hrs.replace('!t', hours);
            }
        } 
        
        // days
        else if (day <= 7) {
            if (day == 1)
                return langs[lang].ago.yesterday;
            else{
                // for arabic
                if (lang == 'ar' && day >= 11) {
                    return langs[lang].ago.days.single.replace('!t', day);
                } else if (lang == 'ar' && day <= 11) {
                    return langs[lang].ago.days.many.replace('!t', day);
                }
                // else
                return langs[lang].ago.days.replace('!t', day);
            }
        } 

        // weeks
        else if (week <= 4.3) {
            if (week == 1)
                return langs[lang].ago.oneweek;
            else{
                // for arabic
                if (lang == 'ar' && week >= 11) {
                    return langs[lang].ago.weeks.single.replace('!t', week);
                } else if (lang == 'ar' && week <= 11) {
                    return langs[lang].ago.weeks.many.replace('!t', week);
                }
                // else
                return langs[lang].ago.weeks.replace('!t', week);
            }
        } 
        
        // months
        else if (month <= 12) {
            if (month == 1)
                return langs[lang].ago.onemonth;
            else{
                // for arabic
                if (lang == 'ar' && month >= 11) {
                    return langs[lang].ago.months.single.replace('!t', month);
                } else if (lang == 'ar' && month <= 11) {
                    return langs[lang].ago.months.many.replace('!t', month);
                }
                // else
                return langs[lang].ago.months.replace('!t', month);
            }
        } 
        
        // years
        else {
            if (year == 1)
                return langs[lang].ago.oneyear;
            else{
                // for arabic
                if (lang == 'ar' && year >= 11) {
                    return langs[lang].ago.years.single.replace('!t', year);
                } else if (lang == 'ar' && year <= 11) {
                    return langs[lang].ago.years.many.replace('!t', year);
                }
                // else
                return langs[lang].ago.years.replace('!t', year);
            }
        }

    }

    // Main function
    // nta (time-ago) function
    function nta(date, lang, format) {
        if (!date) {
            throw_error('Please give the date as the first parameter');
        }

        // set default language to arabic
        var lang = lang ? lang : 'ar';

        if(format === 'full')
            return fullFormat(date, lang)
        
        // else
        return agoFormat(date, lang)

    }


    // return the function
    return nta;

}));